File: miscellaneous/ltcd_376.py
'''
LeetCode - problem 376
'''

    def wiggleMaxLength(self, nums: List[int]) -> int:
        memo={}
        def recurse(index,state):
            if index==len(nums):
                return 0
            if index in memo:
                return memo[index]
            ans=1
            for i in range(index+1,len(nums)):
                if state=='positive' and nums[i]-nums[index]<0:
                    ans=max(ans,1+recurse(i,'negative'))
                elif state=='negative'and nums[i]-nums[index]>0:
                    ans=max(ans,1+recurse(i,'positive'))
            memo[index]=ans
            return memo[index]
        res=float('-inf')
        for i in range(1,len(nums)):
            if nums[i]-nums[i-1]>0:
                res=max(res,1+recurse(i,'positive'))
            elif nums[i]-nums[i-1]<0:
                res=max(res,1+recurse(i,'negative'))
        if res==float('-inf'):
            return 1
        return res

File: miscellaneous/ltcd_376.cpp
//LeetCode - problem 376

    int wiggleMaxLength(vector<int>& a) {
        
        int n = a.size();
        if(n == 0) return 0;
        
        vector<int> up(n,1), down(n,1);
        for(int i = 1 ; i < n ; i++) {
            up[i] = up[i-1], down[i] = down[i-1];
        	if(a[i] > a[i-1]) up[i] = down[i-1] + 1;
        	if(a[i] < a[i-1]) down[i] = up[i-1] + 1;
        }

    	return max(up[n-1], down[n-1]);
    }
